helpers/helpersImpl.go

```diff
@@ -49,3 +61,118 @@ func (helper *HelperImpl) GetObjects(bucketName string, objects []helperModels.G
     transceiver := newGetTransceiver(bucketName, &objects, &strategy, helper.client)
     return transceiver.transfer()
 }
+
+func (helper *HelperImpl) GetObjectsSpanningJobs(bucketName string, objects []helperModels.GetObject, strategy ReadTransferStrategy) ([]string, error) {
+    // Attempt to send the entire job at once
+    jobId, err := helper.GetObjects(bucketName, objects, strategy)
+    if err == nil {
+        // success
+        return []string{jobId}, nil
+    } else if !helper.isCannotPreAllocateError(err) {
+        // error not related to pre-allocation
+        return nil, err
+    }
+
+    // Retrieve each file individually
+    var jobIds []string
+    for _, getObject := range objects {
+        fileJobIds := helper.retrieveIndividualFile(bucketName, getObject, strategy)
+        jobIds = append(jobIds, fileJobIds...)
+    }
+    return jobIds, nil
+}
+
+func (helper *HelperImpl) isCannotPreAllocateError(err error) bool {
+    badStatusErr, ok := err.(*models.BadStatusCodeError)
+    if !ok || badStatusErr.ActualStatusCode != http.StatusServiceUnavailable {
+        // failed to create bulk get for reason other than 503
+        return false
+    }
+
+    if strings.Contains(badStatusErr.Error(), "GET jobs that have a chunkClientProcessingOrderGuarantee of IN_ORDER must be entirely pre-allocated.  Cannot pre-allocate") {
+        return true
+    }
+    return false
+}
+
+func (helper *HelperImpl) retrieveIndividualFile(bucketName string, getObject helperModels.GetObject, strategy ReadTransferStrategy) []string {
+    // Get the blob offsets
+    headObject, err := helper.client.HeadObject(models.NewHeadObjectRequest(bucketName, getObject.Name))
+    if err != nil {
+        getObject.ChannelBuilder.SetFatalError(err)
+        return nil
+    }
+    var offsets []int64
+    for offset := range headObject.BlobChecksums {
+        offsets = append(offsets, offset)
+    }
+
+    sort.Slice(offsets, func(i, j int) bool {
+        return offsets[i] < offsets[j]
+    })
+
+    // Get the object size
+    objectsDetails, err := helper.client.GetObjectsWithFullDetailsSpectraS3(
+        models.NewGetObjectsWithFullDetailsSpectraS3Request().
+            WithBucketId(bucketName).WithName(getObject.Name).
+            WithLatest(true))
+
+    if err != nil {
+        getObject.ChannelBuilder.SetFatalError(err)
+        return nil
+    } else if len(objectsDetails.DetailedS3ObjectList.DetailedS3Objects) < 1 {
+        getObject.ChannelBuilder.SetFatalError(fmt.Errorf("failed to get object details"))
+        return nil
+    }
+
+    // Retrieve the object one blob at a time in order
```

Comment:
Is it realistic for the blob sizes to be bigger than cache? I guess they would have to intentionally shrink the cache after the put to get that to happen.

Reply:
I was testing on the BP Sim and when I put an object when cache was 18 GB, the blobs were 10 MB. The min cache size is 1 GB. So I think we're good.